com.unity.sg2/Editor/GraphUI/GraphElements/ModelUI/Inspector/SubgraphOutputsInspector.cs

```diff
@@ -0,0 +1,157 @@
+                makeOptionsUnique: false,
+                makeListReorderable: false
+            );
+
+            m_OutputRowListField.listView.name = "sg-subgraph-output-list";
+
+            var controller = new SubgraphOutputListViewController();
+            controller.beforeItemsRemoved += indices =>
+            {
+                foreach (var i in indices)
+                {
+                    m_OwnerElement.RootView.Dispatch(new RemoveContextEntryCommand((GraphDataContextNodeModel)model, m_OutputRows[i].Name));
+                }
+            };
+
+            m_OutputRowListField.listView.SetViewController(controller);
+
+            m_OutputRowListField.listView.makeItem = () =>
+            {
+                var ve = new VisualElement();
+                // TODO (Joe): SG1's fields use the style of labels, with no field background.
+                GraphElementHelper.LoadTemplateAndStylesheet(ve, "SubgraphOutputRow", "sg-subgraph-output-row");
+                ve.Q<DropdownField>().choices = typeNames;
+                return ve;
+            };
+
+            m_OutputRowListField.listView.bindItem = (ve, i) =>
+            {
+                var textField = ve.Q<TextField>();
+                textField.SetValueWithoutNotify(m_OutputRows[i].Name);
+                textField.isDelayed = true;
+                textField.RegisterValueChangedCallback(evt =>
+                {
+                    if (evt.newValue == evt.previousValue) return;
+                    var uniqueName = GetUniqueOutputName(evt.newValue);
+                    m_OwnerElement.RootView.Dispatch(new RenameContextEntryCommand(contextNodeModel, m_OutputRows[i].Name, uniqueName));
+                });
+
+                var dropdownField = ve.Q<DropdownField>();
+                dropdownField.SetValueWithoutNotify(dropdownField.choices[m_OutputRows[i].TypeIndex]);
+                dropdownField.RegisterValueChangedCallback(_ => // Event gives us a string, not the index
+                {
+                    m_OwnerElement.RootView.Dispatch(new ChangeContextEntryTypeCommand(contextNodeModel, m_OutputRows[i].Name, m_AvailableTypes[dropdownField.index]));
+                });
+            };
+        }
+
+        string GetTypeDisplayName(TypeHandle typeHandle)
+        {
+            return m_Stencil.TypeMetadataResolver.Resolve(typeHandle)?.FriendlyName ?? typeHandle.Name;
+        }
+
+        string GetUniqueOutputName(string proposedName)
+        {
+            // Make sure this name can't be interpreted as a path
+            // TODO (Joe): It'd be better to separate the display and data names
+            proposedName = proposedName.Replace('.', '_');
+
+            if (!m_ContextNodeModel.TryGetNodeReader(out var nodeHandler)) return proposedName;
+
+            var existingNames = nodeHandler.GetPorts().Select(p => p.ID.LocalPath);
+            return ObjectNames.GetUniqueName(existingNames.ToArray(), proposedName);
```

Comment:
Oooo I didn't know this exists, should be handy for the future, we might need it for blackboard item names